src/gluonts/dataset/dataframe.py

```diff
@@ -0,0 +1,300 @@
+        elements are removed when iterating over the data set. This becomes
+        important when the predictor is called.
+    """
+
+    dataframes: Union[
+        List[pd.DataFrame], Dict[str, pd.DataFrame], pd.DataFrame
+    ]
+    target: Union[str, List[str]]
+    timestamp: str
+    freq: str
+    feat_dynamic_real: List[str] = field(default_factory=list)
+    feat_dynamic_cat: List[str] = field(default_factory=list)
+    feat_static_real: List[str] = field(default_factory=list)
+    feat_static_cat: List[str] = field(default_factory=list)
+    past_feat_dynamic_real: List[str] = field(default_factory=list)
+    ignore_last_n_targets: int = 0
+
+    def __post_init__(self) -> None:
+        if isinstance(self.target, list) and len(self.target) == 1:
+            self.target = self.target[0]
+        self.one_dim_target = not isinstance(self.target, list)
+        self.process = ProcessDataEntry(
+            self.freq, one_dim_target=self.one_dim_target
+        )
+        # store data internally as List[Tuple[str, pandas.DataFrame]]
+        # if str is not empty it will be set in ``DataEntry`` as ``item_id``.
+        if isinstance(self.dataframes, dict):
+            self._dataframes = list(self.dataframes.items())
+        elif isinstance(self.dataframes, list):
+            self._dataframes = [("", df) for df in self.dataframes]
+        else:  # case single dataframe
+            self._dataframes = [("", self.dataframes)]
+
+    def _dataentry(self, item_id: str, df: pd.DataFrame) -> DataEntry:
+        assert check_timestamps(
+            df.loc[:, self.timestamp].to_list(), freq=self.freq
+        ), "``timestamps`` are not monotonically increasing or evenly spaced."
+        dataentry = as_dataentry(
+            data=df,
+            target=self.target,
+            timestamp=self.timestamp,
+            feat_dynamic_real=self.feat_dynamic_real,
+            feat_dynamic_cat=self.feat_dynamic_cat,
+            feat_static_real=self.feat_static_real,
+            feat_static_cat=self.feat_static_cat,
+            past_feat_dynamic_real=self.past_feat_dynamic_real,
+        )
+        if item_id:
+            dataentry["item_id"] = item_id
+        return dataentry
+
+    def __iter__(self) -> Iterator[DataEntry]:
+        for item_id, df in self._dataframes:
+            dataentry = self.process(self._dataentry(item_id, df))
+            if self.ignore_last_n_targets:
+                dataentry = prepare_prediction_data(
+                    dataentry, self.ignore_last_n_targets
+                )
+            yield dataentry
+
+    def __len__(self) -> int:
+        return len(self._dataframes)
+
+
+@dataclass
+class LongDataFrameDataset(Dataset):
+    """
+    This dataset uses the long format for each variable. Target time series
+    values, for example, are stacked on top of each other rather than
+    side-by-side. The same is true for other dynamic or categorical features.
```

Comment:
I think @jaheba meant to use an alternative constructor for the class above. Something like

```python
@classmethod
def from_long_dataframe(cls, df: pd.DataFrame, item_id: str, **kwargs) -> "DataFramesDataset":
    return cls(dataframes=dict(list(df.groupby(item_id))), **kwargs)
```

should allow saving the definition of this class entirely